File: java/src/test/java/com/penapereira/patterns/objectpool/ObjectPoolTest.java
package com.penapereira.patterns.objectpool;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotSame;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.Test;

class ObjectPoolTest {

    @Test
    void createsObjectsLazilyUpToCapacity() {
        ObjectPool pool = new ObjectPool(2);
        assertEquals(0, pool.created());
        Reusable first = pool.acquire("a");
        Reusable second = pool.acquire("b");
        assertNotSame(first, second);
        assertEquals(1, first.id());
        assertEquals(2, second.id());
        assertEquals(2, pool.created());
        assertEquals(2, pool.inUse());
        assertEquals(0, pool.available());
    }

    @Test
    void refusesWhenExhausted() {
        ObjectPool pool = new ObjectPool(1);
        pool.acquire("a");
        IllegalStateException e = assertThrows(IllegalStateException.class, () -> pool.acquire("b"));
        assertEquals("pool exhausted, 1 of 1 in use", e.getMessage());
    }

    @Test
    void reusesReleasedObjectsInFifoOrder() {
        ObjectPool pool = new ObjectPool(3);
        Reusable first = pool.acquire("a");
        Reusable second = pool.acquire("b");
        pool.release(second);
        pool.release(first);
        assertSame(second, pool.acquire("c"));
        assertSame(first, pool.acquire("d"));
        assertEquals(2, pool.created());
        assertEquals(2, first.uses());
    }

    @Test
    void releaseResetsTheObject() {
        ObjectPool pool = new ObjectPool(1);
        Reusable reusable = pool.acquire("a");
        assertEquals("a", reusable.task());
        pool.release(reusable);
        assertNull(reusable.task());
        assertEquals(1, pool.available());
        assertEquals(0, pool.inUse());
    }

    @Test
    void rejectsReleasingAnObjectThatIsNotInUse() {
        ObjectPool pool = new ObjectPool(1);
        Reusable reusable = pool.acquire("a");
        pool.release(reusable);
        IllegalArgumentException e = assertThrows(IllegalArgumentException.class, () -> pool.release(reusable));
        assertEquals("Reusable#1 is not in use", e.getMessage());
    }

    @Test
    void rejectsANonPositiveCapacity() {
        assertThrows(IllegalArgumentException.class, () -> new ObjectPool(0));
    }
}

File: javascript/src/object-pool/object-pool.test.js
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import { ObjectPool } from "./object-pool.js";

describe("ObjectPool", () => {
  it("creates objects lazily up to capacity", () => {
    const pool = new ObjectPool(2);
    assert.equal(pool.created, 0);
    const first = pool.acquire("a");
    const second = pool.acquire("b");
    assert.notEqual(first, second);
    assert.deepEqual([first.id, second.id], [1, 2]);
    assert.deepEqual([pool.created, pool.inUse, pool.available], [2, 2, 0]);
  });

  it("refuses when exhausted", () => {
    const pool = new ObjectPool(1);
    pool.acquire("a");
    assert.throws(() => pool.acquire("b"), { message: "pool exhausted, 1 of 1 in use" });
  });

  it("reuses released objects in FIFO order", () => {
    const pool = new ObjectPool(3);
    const first = pool.acquire("a");
    const second = pool.acquire("b");
    pool.release(second);
    pool.release(first);
    assert.equal(pool.acquire("c"), second);
    assert.equal(pool.acquire("d"), first);
    assert.equal(pool.created, 2);
    assert.equal(first.uses, 2);
  });

  it("release resets the object", () => {
    const pool = new ObjectPool(1);
    const reusable = pool.acquire("a");
    assert.equal(reusable.task, "a");
    pool.release(reusable);
    assert.equal(reusable.task, null);
    assert.deepEqual([pool.available, pool.inUse], [1, 0]);
  });

  it("rejects releasing an object that is not in use", () => {
    const pool = new ObjectPool(1);
    const reusable = pool.acquire("a");
    pool.release(reusable);
    assert.throws(() => pool.release(reusable), { message: "Reusable#1 is not in use" });
  });

  it("rejects a non-positive capacity", () => {
    assert.throws(() => new ObjectPool(0), { message: "capacity must be positive: 0" });
    assert.equal(new ObjectPool(4).capacity, 4);
  });
});
